File: src/main/java/com/lessmarkup/userinterface/nodehandlers/configuration/NodeListNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.configuration;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.recordmodel.RecordModelCache;
import com.lessmarkup.interfaces.structure.ChildHandlerSettings;
import com.lessmarkup.interfaces.structure.ConfigurationHandler;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;
import com.lessmarkup.userinterface.model.configuration.NodeSettingsModel;

import java.util.OptionalLong;

@ConfigurationHandler(titleTextId = TextIds.VIEWS_TREE)
public class NodeListNodeHandler extends AbstractNodeHandler {
    public static String getHandlerName(Class<?> handlerType, String moduleType) {
        String typeName = handlerType.getName();
        if (typeName.endsWith("NodeHandler")) {
            typeName = typeName.substring(0, typeName.length() - "NodeHandler".length());
        }
        return typeName + " / " + moduleType;
    }

    public static class LayoutInfo {
        private long nodeId;
        private int level;

        public long getNodeId() {
            return nodeId;
        }

        public void setNodeId(long nodeId) {
            this.nodeId = nodeId;
        }

        public int getLevel() {
            return level;
        }

        public void setLevel(int level) {
            this.level = level;
        }
    }

    private final ModuleProvider moduleProvider;
    private final DataCache dataCache;

    @Inject
    public NodeListNodeHandler(ModuleProvider moduleProvider, DataCache dataCache) {
        this.moduleProvider = moduleProvider;
        this.dataCache = dataCache;
        addScript("scripts/controllers/NodeListController");
    }

    @Override
    public JsonObject getViewData() {
        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);

        NodeSettingsModel node = DependencyResolver.resolve(NodeSettingsModel.class);

        JsonObject ret = new JsonObject();

        NodeSettingsModel rootNode = node.getRootNode();

        ret.add("root", rootNode != null ? JsonSerializer.serializePojoToTree(node.getRootNode()) : JsonNull.INSTANCE);
        ret.addProperty("nodeSettingsModelId", modelCache.getDefinition(NodeSettingsModel.class).getId());

        JsonArray handlers = new JsonArray();
        moduleProvider.getNodeHandlers().forEach(id -> {
            JsonObject o = new JsonObject();
            o.addProperty("id", id);
            Tuple<Class<? extends NodeHandler>, String> handler = moduleProvider.getNodeHandler(id);
            o.addProperty("name", getHandlerName(handler.getValue1(), handler.getValue2()));
            handlers.add(o);
        });
        ret.add("nodeHandlers", handlers);
        return ret;
    }

    public JsonElement updateParent(long nodeId, OptionalLong parentId, int order) {
        NodeSettingsModel node = DependencyResolver.resolve(NodeSettingsModel.class);
        node.setNodeId(nodeId);
        node.setParentId(parentId);
        node.setPosition(order);
        return node.updateParent();
    }

    public Object createNode(NodeSettingsModel node) {
        return node.createNode();
    }

    public Object deleteNode(long id) {
        NodeSettingsModel node = DependencyResolver.resolve(NodeSettingsModel.class);
        node.setNodeId(id);
        return node.deleteNode();
    }

    public Object updateNode(NodeSettingsModel node) {
        return node.updateNode();
    }

    public Object changeSettings(long nodeId, JsonElement settings) {
        NodeSettingsModel node = DependencyResolver.resolve(NodeSettingsModel.class);
        node.setSettings(settings);
        node.setNodeId(nodeId);
        return node.changeSettings();
    }

    @Override
    public boolean hasChildren() {
        return true;
    }

    @Override
    public ChildHandlerSettings getChildHandler(String path) {
        String[] split = path.split("/");
        long nodeId;
        if (split.length != 2 || !"access".equals(split[1])) {
            return null;
        }
        try {
            nodeId = Long.parseLong(split[0]);
        } catch (Exception e) {
            return null;
        }
        NodeAccessNodeHandler handler = DependencyResolver.resolve(NodeAccessNodeHandler.class);
        handler.initialize(nodeId);
        handler.initialize(OptionalLong.of(nodeId), null, path, getFullPath() + "/" + path, getAccessType());

        ChildHandlerSettings ret = new ChildHandlerSettings();
        ret.setHandler(handler);
        ret.setPath(path);
        ret.setTitle(LanguageHelper.getFullTextId(Constants.ModuleTypeMain(), TextIds.NODE_ACCESS));
        ret.setId(OptionalLong.of(nodeId));
        return ret;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/configuration/NodeListNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.configuration

import com.google.gson.{JsonArray, JsonElement, JsonNull, JsonObject}
import com.google.inject.Inject
import com.lessmarkup.framework.helpers.{DependencyResolver, JsonSerializer, LanguageHelper}
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.interfaces.annotations.ConfigurationHandler
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.module.ModuleProvider
import com.lessmarkup.interfaces.recordmodel.RecordModelCache
import com.lessmarkup.interfaces.structure.{NodeHandler, NodeHandlerFactory, ChildHandlerSettings}
import com.lessmarkup.userinterface.model.configuration.NodeSettingsModel
import com.lessmarkup.{Constants, TextIds}

object NodeListNodeHandler {

  def getHandlerName(handlerType: Class[_], moduleType: String): String = {
    var typeName: String = handlerType.getName
    if (typeName.endsWith("NodeHandler")) {
      typeName = typeName.substring(0, typeName.length - "NodeHandler".length)
    }
    typeName + " / " + moduleType
  }

  class LayoutInfo(val nodeId: Long, val level: Int)
}

class NodeListNodeHandlerFactory @Inject() (moduleProvider: ModuleProvider, dataCache: DataCache) extends NodeHandlerFactory {
  override def createNodeHandler(nodeHandlerConfiguration: NodeHandlerConfiguration, arguments: Any*): NodeHandler = {
    new NodeListNodeHandler(moduleProvider, dataCache, nodeHandlerConfiguration)
  }
}

@ConfigurationHandler(titleTextId = TextIds.VIEWS_TREE)
class NodeListNodeHandler(moduleProvider: ModuleProvider, dataCache: DataCache, configuration: NodeHandlerConfiguration) extends AbstractNodeHandler(configuration) {

  override def getScripts: Seq[String] = {
    super.getScripts :+ "scripts/controllers/NodeListController"
  }

  override def getViewData: Option[JsonObject] = {
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val node: NodeSettingsModel = DependencyResolver(classOf[NodeSettingsModel])
    val ret: JsonObject = new JsonObject
    val rootNode: NodeSettingsModel = node.getRootNode
    ret.add("root", if (rootNode != null) JsonSerializer.serializePojoToTree(node.getRootNode) else JsonNull.INSTANCE)
    ret.addProperty("nodeSettingsModelId", modelCache.getDefinition(classOf[NodeSettingsModel]).get.getId)
    val handlers: JsonArray = new JsonArray

    for (id <- moduleProvider.getNodeHandlers) {
      val o = new JsonObject
      o.addProperty("id", id)
      val handler = moduleProvider.getNodeHandler(id).get
      o.addProperty("name", NodeListNodeHandler.getHandlerName(handler._1, handler._2))
      handlers.add(o)
    }

    ret.add("nodeHandlers", handlers)
    Option(ret)
  }

  def updateParent(nodeId: Long, parentId: Option[Long], order: Int): JsonElement = {
    val node: NodeSettingsModel = DependencyResolver(classOf[NodeSettingsModel])
    node.nodeId = nodeId
    node.parentId = parentId
    node.position = order
    node.updateParent()
  }

  def createNode(node: NodeSettingsModel): AnyRef = {
    node.createNode
  }

  def deleteNode(id: Long): AnyRef = {
    val node: NodeSettingsModel = DependencyResolver(classOf[NodeSettingsModel])
    node.nodeId = id
    node.deleteNode()
  }

  def updateNode(node: NodeSettingsModel): AnyRef = {
    node.updateNode()
  }

  def changeSettings(nodeId: Long, settings: JsonElement): AnyRef = {
    val node: NodeSettingsModel = DependencyResolver(classOf[NodeSettingsModel])
    node.settings = settings
    node.nodeId = nodeId
    node.changeSettings
  }

  override def hasChildren: Boolean = true

  override def createChildHandler(path: Seq[String]): Option[ChildHandlerSettings] = {
    if (path.length != 2 || "access" != path(1)) {
      return None
    }

    val nodeId = path.head.toLong

    val flatPath = path.mkString("/")

    val nodeHandlerConfiguration = new NodeHandlerConfiguration(
      objectId = Option(nodeId),
      settings = None,
      accessType = getAccessType,
      path = flatPath,
      fullPath = getFullPath + "/" + flatPath
    )

    val handler = createChildHandler(classOf[NodeAccessNodeHandlerFactory], nodeHandlerConfiguration)

    Option(new ChildHandlerSettings(
      handler = handler,
      id = nodeHandlerConfiguration.objectId,
      title = LanguageHelper.getFullTextId(Constants.ModuleTypeMain, TextIds.NODE_ACCESS),
      path = flatPath,
      rest = Nil
    ))
  }
}